sdk/communication/azure-communication-callingserver/src/main/java/com/azure/communication/callingserver/ContentDownloader.java

```diff
@@ -0,0 +1,311 @@
+                                Mono.error(new IllegalStateException("Blob was modified mid download. It was "
+                                    + "originally 0 bytes and is now larger."));
+                            }
+                            return Mono.zip(Mono.just(0L), Mono.just(response));
+                        });
+                }
+                return Mono.error(exception);
+            });
+    }
+
+    private long extractTotalBlobLength(String contentRange) {
+        return contentRange == null ? 0 : Long.parseLong(contentRange.split("/")[1]);
+    }
+
+    private int calculateNumBlocks(long dataSize, long blockLength) {
+        // Can successfully cast to an int because MaxBlockSize is an int, which this expression must be less than.
+        int numBlocks = toIntExact(dataSize / blockLength);
+        // Include an extra block for trailing data.
+        if (dataSize % blockLength != 0) {
+            numBlocks++;
+        }
+        return numBlocks;
+    }
+
+    private <T> Flux<T> downloadChunk(Integer chunkNum, Response<Flux<ByteBuffer>> initialResponse, HttpRange finalRange,
+                                      ParallelDownloadOptions parallelDownloadOptions, long newCount,
+                                      Function<HttpRange, Mono<Response<Flux<ByteBuffer>>>> downloader,
+                                      Function<Response<Flux<ByteBuffer>>, Flux<T>> returnTransformer) {
+        if (chunkNum == 0) {
+            return returnTransformer.apply(initialResponse);
+        }
+
+        // Calculate whether we need a full chunk or something smaller because we are at the end.
+        long modifier = chunkNum.longValue() * parallelDownloadOptions.getBlockSizeLong();
+        long chunkSizeActual = Math.min(parallelDownloadOptions.getBlockSizeLong(),
+            newCount - modifier);
+        HttpRange chunkRange = new HttpRange(finalRange.getOffset() + modifier, chunkSizeActual);
+
+        // Make the download call.
+        return downloader.apply(chunkRange)
+            .subscribeOn(Schedulers.boundedElastic())
+            .flatMapMany(returnTransformer);
+    }
+
+    private static Mono<Void> writeBodyToFile(Response<Flux<ByteBuffer>> response, AsynchronousFileChannel file,
+                                              long chunkNum, ParallelDownloadOptions parallelDownloadOptions,
+                                              Lock progressLock,
+                                              AtomicLong totalProgress) {
+        // Extract the body.
+        Flux<ByteBuffer> data = response.getValue();
+
+        // Report progress as necessary.
+        data = ProgressReporter.addParallelProgressReporting(data,
+            parallelDownloadOptions.getProgressReceiver(), progressLock,
+            totalProgress);
+
+        // Write to the file.
+        return FluxUtil.writeFile(data, file, chunkNum * parallelDownloadOptions.getBlockSizeLong());
```

Comment:
Given that there is an initial download block size and a block size will this reliably write to the proper offset?

Reply:
I left it as is, given that I removed the `initialBlockSize` variable.